### PR TITLE
[VL] Align whole stage pipeline metric with Spark

### PR DESCRIPTION
**Pull request by Yohahaha**
Spark measure pipeline time since the iterator was created, but Gluten measure it in each call.

```
  /**
   * Returns the elapsed time since this object is created. This object represents a pipeline so
   * this is a measure of how long the pipeline has been running.
   */
  public long durationMs() {
    return (System.nanoTime() - startTimeNs) / (1000 * 1000);
  }
```

Following patch will add another metric to reflect native task execution time.

### CODE REVIEW

**Comment by zhztheplayer** [comment]
Any visible difference on the pipeline metrics after making this change?

@rui-mo

**Comment by Yohahaha** [comment]
> Any visible difference on the pipeline metrics after making this change?
> 
> cc @rui-mo

before it record elapsed time of `hasNext` and `next`, after it record from iterator created to last `hasNext` align with Spark's behavior.

**Comment by zhztheplayer** [comment]
> > Any visible difference on the pipeline metrics after making this change?
> > cc @rui-mo
> 
> before it record elapsed time of `hasNext` and `next`, after it record from iterator created to last `hasNext` align with Spark's behavior.

OK. So did you observe any difference on the metric numbers when running your queries from Spark GUI?

**Comment by Yohahaha** [comment]
> > > Any visible difference on the pipeline metrics after making this change?
> > > cc @rui-mo
> > 
> > 
> > before it record elapsed time of `hasNext` and `next`, after it record from iterator created to last `hasNext` align with Spark's behavior.
> 
> OK. So did you observe any difference on the metric numbers when running your queries from Spark GUI?

Yes, sometimes stage UI show no difference between Gluten and Vanilla, but pipeline metric show Gluten has benefits. We want align metric between Gluten and Vanilla as much as possible, then find which part is slower.

**Comment by zzcclp** [comment]
It seems that the changes do not impact the ch backend. `IteratorCompleter` is only for the velox backend.

**Comment by rui-mo** [comment]
> before it record elapsed time of hasNext and next, after it record from iterator created to last hasNext align with Spark's behavior

@Yohahaha Compared with previous logic, looks more time is recorded in this metric, right?

**Comment by Yohahaha** [comment]
> > before it record elapsed time of hasNext and next, after it record from iterator created to last hasNext align with Spark's behavior
> 
> @Yohahaha Compared with previous logic, looks more time is recorded in this metric, right?

Yes, previous logic more like record velox task execution time, but we may not ignore other parts' time.

**Review by rui-mo — APPROVED** [review]

**Comment by Yohahaha** [comment]
could we merge?

**Comment by Yohahaha** [comment]
@PHILO-HE @ulysses-you could you help merge? thanks!

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala [hunk 1]
```diff
@@ -184,8 +179,6 @@ class IteratorApiImpl extends IteratorApi with Logging {
         rootNode.toProtobuf.toByteArray,
         columnarNativeIterator)
 
-    pipelineTime += TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - beforeBuild)
```
Comment (ulysses-you) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:187]: thanky you @Yohahaha . Shall we add one more metrics for the durationMs of preparing the native input iterator ? So the previous pipeline time can be broken into two parts, one is for preparing and the other is for executing.

Comment (Yohahaha) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:187]: ok, I can open an issue to track it.

**Comment by ulysses-you** [comment]
this pr does not affect CH backend, FWI @zzcclp

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala (hunk 1 above)
Comment (Yohahaha) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:187]: https://github.com/oap-project/gluten/issues/3830

**Pull request merged**